### ISSUE TITLE
Rule: Prefer using ViewModels instead of extending Block classes

### ISSUE THREAD

**Issue by jahvi** [extdn/extdn-phpcs#31]
Unless we're overriding core functionality I don't think extending the Magento `Template` or `AbstractBlock` classes is needed anymore, in fact it can lead to messy block classes by having to extend the parent block's constructor properties encouraging code duplication.

**Comment by jissereitsma** [comment]
Completely agree. We have a couple of these things that lead to Magento version-specific rules though. ViewModels can only be used from 2.2 onwards. PHP 7.1 syntax like-wise. But I would love this to be shipped in a rule myself.

**Comment by jahvi** [comment]
Yes not sure what the plan is to support multiple Magento and PHP versions
for this library but it would be a nice addition.

**Comment by schmengler** [comment]
I would like this code sniffer to be oriented towards the future, not the past. If the need for a 2.1 compatibility branch (soon: 2.2 compatibility branch) arises, and people are willing to work on that, fine. But let's embrace the standards and possibilities of the latest version.

Those Magento version issues aside, I partially agree: Extending `Template` directly should never be necessary and would make a good code sniffer rule. If we have to work with legacy blocks from the core, we can still extend those. *But* extending `AbstractBlock` *without template* should be allowed, for special cases with custom rendering without a template.

**Comment by jahvi** [comment]
Agree I'd also support this library focusing on 2.2+.

What would be the use case for extending the `AbstractBlock` though? I might be missing something here but even if you don't need a template wouldn't using a `ViewModel` still accomplish the same thing?

**Comment by schmengler** [comment]
@jahvi a use case could be rendering of inline SVG, or anything that does not use a PHTML template to generate content.

it would be possible with a template that just contains something like `<?= $block->getViewModel()->toHtml()`. For me this looks like a workaround, but if we can get rid of unnecessary inheritance, I can live with it.

**Comment by davidverholen** [comment]
@schmengler  would it make sense to disallow any public functions (except overridden from AbstractBlock) in custom blocks?

You still could have custom blocks and change e.g. _toHtml(), but for any functionality that should be accessible from outside (apart from magento rendering logic) viewmodels should do the trick

**Comment by schmengler** [comment]
@davidverholen smart idea! I can't think of any reason against it

**Comment by navarr** [comment]
Does anyone know if ViewModels are checked for implementing IdentityInterface - or does that sort of thing still require making a custom block

**Comment by davidverholen** [comment]
good point, I would guess that they are not checked. Although it would be possible to check all block arguments for the IdentityInterface. So you could probably introduce a new "default" block (but then it has to be defined in the xml layout block directive) which implements the IdentityInterface and then in the getIdentities method iterates over its own arguments searching for IdentityInterface.

It might be a good Idea to have this in the core? Maybe even in the default Template block?

**Comment by jissereitsma** [comment]
I just checked the code and they are not checked as far as I can see. Indeed as @davidverholen says, this might be a nice idea for a PR to the core.
